app/src/org/commcare/tasks/DataPullTask.java

```diff
@@ -102,258 +101,296 @@ public DataPullTask(String username, String password,
+        } else if (pullResponse.responseCode == 500) {
+            return handleServerError();
+        }
+        return null;
+    }
+
+    private ResultAndError<PullTaskResult> handleAuthFailed() {
+        wipeLoginIfItOccurred();
+        Logger.log(AndroidLogger.TYPE_USER, "Bad Auth Request for user!|" + username);
+        return new ResultAndError<>(PullTaskResult.AUTH_FAILED);
+    }
+
+    /**
+     * @return the proper result, or null if we have not yet been able to determine the result to
+     * return
+     * @throws IOException
+     */
+    private ResultAndError<PullTaskResult> handleSuccessResponseCode(
+            RemoteDataPullResponse pullResponse, AndroidTransactionParserFactory factory)
+            throws IOException {
+
+        handleLoginNeededOnSuccess();
+        this.publishProgress(PROGRESS_AUTHED, 0);
+
+        if (isCancelled()) {
+            // About to enter data commit phase; last chance to finish early if cancelled.
+            return new ResultAndError<>(PullTaskResult.UNKNOWN_FAILURE, "");
+        }
+
+        this.publishProgress(PROGRESS_DOWNLOADING_COMPLETE, 0);
+        Logger.log(AndroidLogger.TYPE_USER, "Remote Auth Successful|" + username);
+
+        try {
+            BitCache cache = pullResponse.writeResponseToCache(context);
+            String syncToken = readInput(cache.retrieveCache(), factory);
+            updateUserSyncToken(syncToken);
+
+            onSuccessfulSync();
+            return new ResultAndError<>(PullTaskResult.DOWNLOAD_SUCCESS);
+        } catch (XmlPullParserException e) {
+            e.printStackTrace();
+            Logger.log(AndroidLogger.TYPE_USER,
+                    "User Sync failed due to bad payload|" + e.getMessage());
+            return new ResultAndError<>(PullTaskResult.BAD_DATA, e.getMessage());
+        } catch (ActionableInvalidStructureException e) {
+            e.printStackTrace();
+            Logger.log(AndroidLogger.TYPE_USER,
+                    "User Sync failed due to bad payload|" + e.getMessage());
+            return new ResultAndError<>(PullTaskResult.BAD_DATA_REQUIRES_INTERVENTION,
+                    e.getLocalizedMessage());
+        } catch (InvalidStructureException e) {
+            e.printStackTrace();
+            Logger.log(AndroidLogger.TYPE_USER,
+                    "User Sync failed due to bad payload|" + e.getMessage());
+            return new ResultAndError<>(PullTaskResult.BAD_DATA, e.getMessage());
+        } catch (UnfullfilledRequirementsException e) {
+            e.printStackTrace();
+            Logger.log(AndroidLogger.TYPE_ERROR_ASSERTION,
+                    "User sync failed oddly, unfulfilled reqs |" + e.getMessage());
+            return null;
+        } catch (IllegalStateException e) {
+            e.printStackTrace();
+            Logger.log(AndroidLogger.TYPE_ERROR_ASSERTION,
+                    "User sync failed oddly, ISE |" + e.getMessage());
+            return null;

```

Comment:
For the reason I explained above - in the original code (https://github.com/dimagi/commcare-odk/blob/master/app/src/org/commcare/tasks/DataPullTask.java#L278-L284) neither of these blocks return anything, causing them to fall through to the return code at the end. If you think that wasn't intentional or important, I can have them return something here.